### DISCUSSION TITLE
hwrand32() Seed

### DISCUSSION THREAD

**Discussion by NuclearPhoenixx** [earlephilhower/arduino-pico#1008] [Ideas]
Hi there,

I'm building a true random number generator using radioactive decay and a Pico, which obviously is a great source for randomness, but pretty limited in the amount of numbers you can get out of that, of course.
Now I'd like to use the pseudorandom generator with these truly random numbers as seeds, so that I can somewhat combine the advantages of both methods.

Unfortunately, the function `rp2040.hwrand32()` only supports seeds using the CPU cycle counter, if I understood that correctly. How difficult would it be to supply your own seed with this function? I have no idea how you implemented that, but just calling `uint32_t rp2040.hwrand32(uint32_t seed)` or something like that would be awesome. 😃

**Comment by earlephilhower** [comment]
For a HW random generator there's no such thing as a seed, though.  Only PRNGs have them.  So I don't think the proposed function signature makes sense, sorry.

You could take your random source, collect enough bits, and then just `srand(xxx)` and use the standard PRNG generators built in to Newlib, though.  No real reason to use the class after the initial read

**Reply by NuclearPhoenixx** [reply]
Oops! I think I've completely misunderstood then. I thought it was just a PRNG, my fault, sorry.

What source of entropy does the HW RNG use then and why is it neither reliable in a sense of True RNG nor is it a PRNG?

**Reply by earlephilhower** [reply]
There's an internal ring oscillator, documented in the datasheet.  It is free-running and you can read one bit from it.  It's not a perfect random generator, but it's not any mathematical function like a PRNG.

**Reply by NuclearPhoenixx** [reply]
Alright, I see. Thank you so much!
